lib/pure/asyncnet.nim

```diff
@@ -780,6 +780,123 @@ proc isClosed*(socket: AsyncSocket): bool =
   ## Determines whether the socket has been closed.
   return socket.closed
 
+proc sendTo*(socket: AsyncSocket, data: pointer, dataSize: int, address: string,
+             port: Port, flags = {SocketFlag.SafeDisconn}): owned(Future[void])
+            {.async, since: (1, 3).} =
+  ## This proc sends ``data`` to the specified ``address``, which may be an IP
+  ## address or a hostname. If a hostname is specified this function will try
+  ## each IP of that hostname. The returned future will complete once all data
+  ## has been sent.
+  ## 
+  ## If an error occurs an OSError exception will be raised.
+  assert(socket.protocol != IPPROTO_TCP, "Cannot `sendTo` on a TCP socket")
+  assert(not socket.closed, "Cannot `sendTo` on a closed socket")
+
+  let
+    aiList = getAddrInfo(address, port, socket.domain, socket.sockType,
```

Comment:
I guess my main concern is that clients using TCP connections will use this instead of `send`, and thus incur the cost of DNS lookup each time they send data (it wouldn't be immediately obvious either, since the behavior is to ignore the hostname). This is in contrast to the C version, which requires the user to pass in a sockaddr data structure.

Additionally, I also find it conceptually confusing (and this is also a criticism of the original C functions) that `sendTo` works with TCP and UDP, while `recvFrom` works with just UDP.

That being said, I'm not super-familiar with UDP and sendTo/recvFrom, so perhaps there's a reason that TCP works for sending data.

Reply:
So, this DNS query cost, if a hostname is sent, is a cost that the person will have, if they use the less appropriate proc. The assert is there, but I can improve the message, as well as add comments in the proc to use the send() proc instead of sendTo().